man/man3/Info.3.md
# Info

[NAME](#name)\
[SYNOPSIS](#synopsis)\
[SUBROUTINES](#subroutines)\
[check_x32](#check_x32)\
[get_download_info](#get_download_info)\
[get_from_info](#get_from_info)\
[get_orig_build_number](#get_orig_build_number)\
[get_orig_version](#get_orig_version)\
[get_requires](#get_requires)\
[get_reverse_reqs](#get_reverse_reqs)\
[get_sbo_build_number](#get_sbo_build_number)\
[get_sbo_version](#get_sbo_version)\
[parse_info](#parse_info)\
[EXIT CODES](#exit-codes)\
[SEE ALSO](#see-also)\
[AUTHORS](#authors)\
[LICENSE](#license)

------------------------------------------------------------------------

## NAME

SBO::Lib::Info − Utilities to get data from SBo info files.

## SYNOPSIS

    use SBO::Lib::Info qw/ get_reqs /;\
    my \@reqs = \@{ get_requires($sbo) };

## SUBROUTINES

### check_x32

    my $bool = check_x32($location);

check_x32() returns a true value if the SlackBuild in \$location
considers 64−bit builds \"UNTESTED\" or \"UNSUPPORTED\". Otherwise, it
returns a false value.

### get_download_info

    my $downloads = get_download_info(LOCATION =\> $location, X64 =\>
\$x64);\
    my $downloads = get_download_info(LOCATION =\> $location);

get_download_info() reads in an info file from \$location. The flag
\$x64 determines whether the 64−bit download files should be used or
not. \$x64 defaults to a true value if unspecified.

This subroutine returns a hashref where each key is a download link; the
corresponding value is the expected md5sum.

### get_from_info

    my $data = get_from_info(LOCATION =\> $location, GET =\> $key);

get_from_info() retrieves the information under \$key from the info file
in \$location.

### get_orig_build_number

    my $build = get_orig_build_number($sbo);

get_orig_build_number() returns the build number in the SlackBuilds.org
tree for the given \$sbo, calling get_sbo_build_number().

### get_orig_version

    my $ver = get_orig_version($sbo);

get_orig_version() returns the version in the SlackBuilds.org tree for
the given \$sbo, calling get_sbo_version().

### get_requires

    my $reqs = get_requires($sbo);

get_requires() returns the requirements for a given \$sbo.

### get_reverse_reqs

    my %required_by = get_reverse_reqs($slackbuilds);

get_reverse_reqs() takes a list of SlackBuilds and returns a hashref
with reverse dependencies among them. \$slackbuilds should ordinarily be
a list of all installed scripts.

### get_sbo_build_number

    my $build = get_sbo_build_number($location);

get_sbo_build_number() returns the build number found in the SlackBuild
in \$location.

### get_sbo_version

    my $ver = get_sbo_version($location);

get_sbo_version() returns the version found in the info file in
\$location.

### parse_info

    my %parse = parse_info($str);

parse_info() parses the contents of an info file from \$str and returns
a key-value list of all values present. It attempts to repair trailing
whitespace, blank lines, garbage lines and missing quotation marks and
backslashes.

## EXIT CODES

Info.pm subroutines can return the following exit codes:

    \_ERR_USAGE 1 usage errors\
    \_ERR_SCRIPT 2 script or module bug

## SEE ALSO

[SBO::Lib(3)](Lib.3.md), [SBO::Lib::Build(3)](Build.3.md), [SBO::Lib::Download(3)](Download.3.md),
[SBO::Lib::Pkgs(3)](Pkgs.3.md), [SBO::Lib::Pkgs(3)](Pkgs.3.md), [SBO::Lib::Readme(3)](Readme.3.md),
[SBO::Lib::Repo(3)](Repo.3.md), [SBO::Lib::Tree(3)](Tree.3.md), [SBO::Lib::Util(3)](Util.3.md)

## AUTHORS

SBO::Lib was originally written by Jacob Pipkin \<j (at) dawnrazor (dot)
net\> with contributions from Luke Williams \<xocel (at) iquidus (dot)
org\> and Andreas Guldstrand \<andreas (dot) guldstrand (at) gmail (dot)
com\>.

SBO::Lib is maintained by K. Eugene Carlson \<kvngncrlsn (at) gmail
(dot) com\>.

## LICENSE

The sbotools are licensed under the MIT License.

Copyright (C) 2012−2017, Jacob Pipkin, Luke Williams, Andreas
Guldstrand.

Copyright (C) 2024−2025, K. Eugene Carlson.

------------------------------------------------------------------------


ChangeLog.md
# ChangeLog
* 1.0 - 2024-11-01
  * Change the git branch to match the Slackware version, or to a user-configured branch
  * Change the default URLs to support Slackware 15.0 and use git repositories
  * Use rsync default mirrors with RSYNC_DEFAULT (except for -current)
  * Detect build number changes and optionally upgrade
  * Detect and report out-of-tree _SBo SlackBuilds with sbocheck
  * Save any build options when running a SlackBuild
  * Install bash completions; zsh completions have been updated
  * Install a default configuration file; tweak sboconfig to play somewhat nicely with it
  * Fix making -compat32 packages for non-i486 SlackBuilds
  * Fix building -compat32 packages from templates

## Historical Changes
* 2.7 - 2019-04-28
  * Actually fix the sbofind -e bug #71
      
* 2.6 - 2019-04-27
  * Compatibility with new perl versions where you need to escape { in regexp
    #75 #77 #78
  * Add a --tries 5 option when downloading from sbosrcarch, which is a
    saner limit than the default of 20. #79
  * Change sboclean options --clean-dist and --clean-work to shorter forms #52
  * Add limited -current support using ponce's SBo repo for -current #73
  * Fix bug with sboinstall --reinstall -r #72
  * Fix bug with sbofind -e #71

* 2.5 - 2018-02-14
  * Document download behaviour #66
  * Remake sbosnap and sboremove to have OO semantics
  * Strip -compat32 from slackbuild names when looking them up #65
  * Optimise searching in sbofind

* 2.4 - 2017-05-18
  * Rewrite sboremove from the ground up so it relies less on global state
  * Fix for parsing README with useradd/groupadd commands which span lines #57
  * Add --reinstall option to sboinstall #58
  * Exit with error when sbosnap fails to sync with a repo #61
  * Add version information to sbofind output #60

* 2.3 - 2017-01-21
  * Bugfix for parsing .info files with \ among the separators #55

* 2.2 - 2017-01-17
  * Bugfix for parsing .info files with trailing whitespace after a value #54

* 2.1 - 2017-01-14
  * Internals:
    - Adding internal documentation
    - Extract code to submodules for easier separation of concerns
  * New features:
    - Support for templates for installing things with specified options #38
    - Display other README files if the slackbuild comes with them #49
  * Bugfixes
    - sboinstall/sboremove disagreeing about a package being installed #44
    - sbocheck and sboupgrade misinterpreting version strings #45
    - parsing .info files without leading space on second line #46
    - local git repo gets partially chowned to root #47
    - stop excluding .tar.gz files when rsyncing #53

* 2.0 - 2016-07-02
  * Major new features
    * LOCAL_OVERRIDES setting

      Allows to keep a directory with local slackbuild dirs that will override
      whatever is found in the regular repository. #8 #13 #14 #15 #19 #20
    * SLACKWARE_VERSION setting

      Allows to specify the slackware version to sync from SBo. Previously only
      the version in your /etc/slackware-version was used for this, and if that
      had gotten updated in -current, you'd have needed to wait both for a new
      version of sbotools, as well as SBo to get the new repository online
      before sbotools would work for you again. #19
    * REPO setting

      This will override the SLACKWARE_VERSION setting. It's used to specify an
      absolute URL for the SBo repository you want to sync with. #6 #19 #27
    * Use sbosrcarch source archive if download fails #7 #19 #24
    * sboupgrade --all option to upgrade everything listed by sbocheck. #9 #19
    * Travis CI integration

      Every push will now cause the test-suite to be run. #18
    * Hundreds of new unit-tests. #18 #19 #23 #24 #25 #27 #28 #31 #32 #33 #35 #41 #43
    * sbofind will now also use tags if they're available #37
  * Minor/bugfixes/documentation fixes
    * Use system perl when running and installing sbotools.
    * sbocheck output changed. #10 #13 #20
    * Better debug messages on errors. #16
    * manpage fixes. #17
    * sboupgrade handles dependencies better. #12 #28
    * Update bundled Sort::Versions to 1.62.
    * sboinstall/upgrade/sbocheck: small bugfixes. #21 #35 #41 #43
    * sbosnap: display download progress, update git trees better. #26 #27

* 1.9 - 2015-11-27
  * Make it compatible with perls newer than 5.18
  * Lots of code cleanup
  * Rewrite build-queue code. #2
  * Fix issue when TMP is set. #4
  * Fix various bugs related to cleanup code
  * Change location of website
  * Fix downloading of multiple sources in newer slackbuilds. #5



SBO-Lib/lib/SBO/Lib/Info.pm
package SBO::Lib::Info;

use 5.016;
use strict;
use warnings;

our $VERSION = '3.4.1';

use SBO::Lib::Util qw/ :const in get_arch get_sbo_from_loc get_optional open_read script_error slurp usage_error uniq wrapsay /;
use SBO::Lib::Tree qw/ get_orig_location get_sbo_location is_local /;

use Exporter 'import';

our @EXPORT_OK = qw{
  check_x32
  get_download_info
  get_from_info
  get_orig_build_number
  get_orig_version
  get_requires
  get_reverse_reqs
  get_sbo_build_number
  get_sbo_version
  parse_info
};

our %EXPORT_TAGS = (
  all => \@EXPORT_OK,
);

=pod

=encoding UTF-8

=head1 NAME

SBO::Lib::Info - Utilities to get data from SBo info files.

=head1 SYNOPSIS

  use SBO::Lib::Info qw/ get_reqs /;

  my @reqs = @{ get_requires($sbo) };

=head1 SUBROUTINES

=cut

=head2 check_x32

  my $bool = check_x32($location);

C<check_x32()> returns a true value if the SlackBuild in C<$location> considers
64-bit builds C<UNTESTED> or C<UNSUPPORTED>. Otherwise, it returns a false value.

=cut

# determine whether or not a given sbo is 32-bit only
sub check_x32 {
  script_error('check_x32 requires an argument.') unless @_ == 1;
  my $dl = get_from_info(LOCATION => shift, GET => 'DOWNLOAD_x86_64');
  return $$dl[0] =~ /UN(SUPPOR|TES)TED/ ? 1 : undef;
}

=head2 get_download_info

  my $downloads = get_download_info(LOCATION => $location, X64 => $x64);
  my $downloads = get_download_info(LOCATION => $location);

C<get_download_info()> reads in an info file from C<$location>. The flag
C<$x64> determines whether the 64-bit download files should be used or not.
C<$x64> defaults to a true value if unspecified.

This subroutine returns a hashref where each key is a download link; the
corresponding value is the expected md5sum.

=cut

# get downloads and md5sums from an sbo's .info file, first
# checking for x86_64-specific info if we are told to
sub get_download_info {
  my %args = (
    LOCATION  => 0,
    X64       => 1,
    @_
  );
  $args{LOCATION} or script_error('get_download_info requires LOCATION.');
  my ($get, $downs, $exit, $md5s, %return);
  $get = ($args{X64} ? 'DOWNLOAD_x86_64' : 'DOWNLOAD');
  $downs = get_from_info(LOCATION => $args{LOCATION}, GET => $get);
  # did we get nothing back, or UNSUPPORTED/UNTESTED?
  if ($args{X64}) {
    if (! $$downs[0] || $$downs[0] =~ qr/^UN(SUPPOR|TES)TED$/) {
      $args{X64} = 0;
      $downs = get_from_info(LOCATION => $args{LOCATION},
        GET => 'DOWNLOAD');
    }
  }
  # if we still don't have any links, something is really wrong.
  return() unless $$downs[0];
  # grab the md5s and build a hash
  $get = $args{X64} ? 'MD5SUM_x86_64' : 'MD5SUM';
  $md5s = get_from_info(LOCATION => $args{LOCATION}, GET => $get);
  return() unless $$md5s[0];
  $return{$$downs[$_]} = $$md5s[$_] for (keys @$downs);
  return \%return;
}

=head2 get_from_info

  my $data = get_from_info(LOCATION => $location, GET => $key);

C<get_from_info()> retrieves the information under C<$key> from the info file
in C<$location>.

=cut

# pull piece(s) of data, GET, from the $sbo.info file under LOCATION.
sub get_from_info {
  my %args = (
    LOCATION  => '',
    GET       => '',
    @_
  );
  unless ($args{LOCATION} && $args{GET}) {
    script_error('get_from_info requires LOCATION and GET.');
  }
  state $store = {LOCATION => ['']};
  my $sbo = get_sbo_from_loc($args{LOCATION});
  return $store->{$args{GET}} if $store->{LOCATION}[0] eq $args{LOCATION};

  # if we're here, we haven't read in the .info file yet.
  my $contents = slurp("$args{LOCATION}/$sbo.info");
  usage_error("get_from_info: could not read $args{LOCATION}/$sbo.info.") unless
    defined $contents;

  my %parse = parse_info($contents);
  script_error("Error when parsing file $sbo.info.") unless %parse;

  $store = {};
  $store->{LOCATION} = [$args{LOCATION}];
  foreach my $k (keys %parse) { $store->{$k} = $parse{$k}; }

  # allow local overrides to get away with not having quite all the fields
  if (is_local($sbo)) {
    for my $key (qw/DOWNLOAD_x86_64 MD5SUM_x86_64 REQUIRES/) {
      $store->{$key} //= ['']; # if they don't exist, treat them as empty
    }
  }
  my @optional = get_optional($sbo);
  if (@optional) {
    for my $requested (@optional) {
      if ($store->{REQUIRES}[0] eq '') {
        $store->{REQUIRES}[0] = $requested;
      } else {
        push @{ $store->{REQUIRES} }, $requested;
      }
    }
  }
  return $store->{$args{GET}};
}

=head2 get_orig_build_number

  my $build = get_orig_build_number($sbo);

C<get_orig_build_number()> returns the build number in the SlackBuilds.org tree for the
given C<$sbo>, calling C<get_sbo_build_number()>.

=cut

sub get_orig_build_number {
  script_error('get_orig_build_number requires an argument.') unless @_ == 1;
  my $sbo = shift;

  my $location = get_orig_location($sbo);

  return $location if not defined $location;

  return get_sbo_build_number($location);
}

=head2 get_orig_version

  my $ver = get_orig_version($sbo);

C<get_orig_version()> returns the version in the SlackBuilds.org tree for the
given C<$sbo>, calling C<get_sbo_version()>.

=cut

sub get_orig_version {
  script_error('get_orig_version requires an argument.') unless @_ == 1;
  my $sbo = shift;

  my $location = get_orig_location($sbo);

  return $location if not defined $location;

  return get_sbo_version($location);
}

=head2 get_requires

  my $reqs = get_requires($sbo);

C<get_requires()> returns the requirements for a given C<$sbo>.

=cut

# wrapper to pull the list of requirements for a given sbo
sub get_requires {
  my $location = get_sbo_location(shift);
  return undef unless $location;
  my $info = get_from_info(LOCATION => $location, GET => 'REQUIRES');
  return $info;
}

=head2 get_reverse_reqs

  my %required_by = get_reverse_reqs($slackbuilds);

C<get_reverse_reqs()> takes a list of SlackBuilds and returns a hashref with
reverse dependencies among them. C<$slackbuilds> should ordinarily
be a list of all installed scripts.

=cut

sub get_reverse_reqs {
  my $slackbuilds = shift;
  my %required_by;

  for my $sbo (keys %$slackbuilds) {
    for my $req (@{ get_requires($sbo) }) {
      $required_by{$req}{$sbo} = 1 if exists $slackbuilds->{$req};
    }
  }

  return \%required_by;
}

=head2 get_sbo_build_number

  my $build = get_sbo_build_number($location);

C<get_sbo_build_number()> returns the build number found in the SlackBuild in
C<$location>.

=cut

# find the build number in the tree for a given sbo (provided a location)
sub get_sbo_build_number {
  script_error('get_sbo_build_number requires an argument.') unless @_ == 1;

  my $location = shift;
  my $sbo = get_sbo_from_loc($location);
  my $build;

  my ($fh, $exit) = open_read("$location/$sbo.SlackBuild");
  usage_error("get_sbo_build_number: could not read $location/$sbo.SlackBuild.") if $exit;

  while (my $line = <$fh>) {
    $build = $line if $line =~ m/^BUILD=/;
    last if $build;
  }
  close $fh;

  $build =~ s/\D//g;

  return $build;
}

=head2 get_sbo_version

  my $ver = get_sbo_version($location);

C<get_sbo_version()> returns the version found in the info file in
C<$location>.

=cut

# find the version in the tree for a given sbo (provided a location)
sub get_sbo_version {
  script_error('get_sbo_version requires an argument.') unless @_ == 1;
  my $version = get_from_info(LOCATION => shift, GET => 'VERSION');
  return $version->[0];
}

=head2 parse_info

  my %parse = parse_info($str);

C<parse_info()> parses the contents of an info file from C<$str> and returns
a key-value list of all values present. It attempts to repair trailing whitespace,
blank lines, garbage lines and missing quotation marks and backslashes.

=cut

sub parse_info {
    script_error('parse_info requires an argument.') unless @_ == 1;
    my $info_str = shift;
    # Fix blank lines
    $info_str =~ s/\n\n/\n/g;
    # Fix trailing whitespace
    $info_str =~ s/\s\n/\n/g;
    # Fix EOF quotation marks
    $info_str =~ s/(?<=[^\"])\n+$/\"\n/g;
    # Fix missing backslashes
    $info_str =~ s/(?<=[^\\\"])\n(?=\s)/\\\n/g;
    # Fix missing terminal quotation marks
    $info_str =~ s/(?<=[^\\\"])\n(?=[A-Z])/\"\n/g;
    # Fix missing initial quotation marks
    my @fields = qw{
        PRGNAM
        VERSION
        HOMEPAGE
        DOWNLOAD
        MD5SUM
        DOWNLOAD_x86_64
        MD5SUM_x86_64
        REQUIRES
        MAINTAINER
        EMAIL
    };
    for my $field (@fields) { $info_str =~ s/(?<=$field)=(?=[^\"])/=\"/g; }
    # Anything that follows a terminal quote and doesn't start KEY="VALUE" is unwanted
    $info_str =~ s/\"\n[^=\"]+(\\|\")\n/\"\n/g;
    # And the start of the file
    $info_str =~ s/^[^=\"]+(\\|\")\n//g;

    my $pos = 0;
    my %ret;

    while ($info_str =~ /\G([A-Za-z0-9_]+)="([^"]*)"\s*(?:\n|\z)/g) {
        my ($key, $val) = ($1, $2);
        $val =~ s/\\[ \t]*$/ /mg;
        my @val = split " ", $val;
        @val = '' unless @val;
        $ret{$key} = \@val;
        $pos = pos($info_str);
    }

    return if $pos != length($info_str);

    return %ret;

}

=head1 EXIT CODES

Info.pm subroutines can return the following exit codes:

  _ERR_USAGE         1   usage errors
  _ERR_SCRIPT        2   script or module bug

=head1 SEE ALSO

SBO::Lib(3), SBO::Lib::Build(3), SBO::Lib::Download(3), SBO::Lib::Pkgs(3), SBO::Lib::Readme(3), SBO::Lib::Repo(3), SBO::Lib::Tree(3), SBO::Lib::Util(3)

=head1 AUTHORS

SBO::Lib was originally written by Jacob Pipkin <j@dawnrazor.net> with
contributions from Luke Williams <xocel@iquidus.org> and Andreas
Guldstrand <andreas.guldstrand@gmail.com>.

=head1 MAINTAINER

SBO::Lib is maintained by K. Eugene Carlson <kvngncrlsn@gmail.com>.

=head1 LICENSE

The sbotools are licensed under the MIT License.

Copyright (C) 2012-2017, Jacob Pipkin, Luke Williams, Andreas Guldstrand.

Copyright (C) 2024-2025, K. Eugene Carlson.

=cut

1;
